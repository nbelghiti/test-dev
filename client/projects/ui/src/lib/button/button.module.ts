import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';

import { IconModule } from '../icon/icon.module';

import { ButtonComponent } from './button/button.component';
import { IconButtonComponent } from './icon-button/icon-button.component';

@NgModule({
	imports: [
		CommonModule,
		IconModule,
		RouterModule,
	],
	declarations: [
		ButtonComponent,
		IconButtonComponent,
	],
	exports: [
		ButtonComponent,
		IconButtonComponent,
	],
})
export class ButtonModule {}
