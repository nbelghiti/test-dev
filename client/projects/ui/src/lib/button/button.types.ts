export enum ButtonType {
	primary = '',
	secondary = 'c-button--secondary',
	tertiary = 'c-button--tertiary',
	destructive = 'c-button--destructive',
}
