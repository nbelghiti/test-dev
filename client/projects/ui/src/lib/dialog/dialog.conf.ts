import { InjectionToken, Type } from '@angular/core';

export const DIALOG_OVERLAY: InjectionToken<Type<unknown>> = new InjectionToken<Type<unknown>>('DIALOG_OVERLAY');
