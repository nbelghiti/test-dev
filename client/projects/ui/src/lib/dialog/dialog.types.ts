export interface IDialogActions {
	[key: string]: (_?) => any;
}

export interface IDialogData {
	[key: string]: any;
}

export interface IDialogResponse {
	message?: string;
	errors?: string[];
}

export interface IDialog {
	closeDialog: () => void;
	completeCallback?: (response?: IDialogResponse) => void;
	actions?: IDialogActions;
	data?: IDialogData;
	hasDescription?: Boolean;
}
