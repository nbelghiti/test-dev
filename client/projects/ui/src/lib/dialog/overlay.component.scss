// =============================================================================
// :: Imports
// =============================================================================
@import "global";

// =============================================================================
// :: Overlay
// =============================================================================
:host {
	position: fixed;
	top: 0;
	left: 0;
	bottom: 0;
	right: 0;
	z-index: $z-index-atmosphere;

	display: flex;
	justify-content: center;
	align-items: center;

	background: $black-50;
}
