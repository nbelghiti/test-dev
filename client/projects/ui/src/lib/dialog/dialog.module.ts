import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { DIALOG_OVERLAY } from './dialog.conf';
import { OverlayComponent } from './overlay.component';

@NgModule({
	imports: [
		CommonModule,
	],
	declarations: [
		OverlayComponent,
	],
	entryComponents: [
		OverlayComponent,
	],
	providers: [{
		provide: DIALOG_OVERLAY,
		useValue: OverlayComponent,
	}],
	exports: [
		OverlayComponent,
	],
})
export class DialogModule { }
