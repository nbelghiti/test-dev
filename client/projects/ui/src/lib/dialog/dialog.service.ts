import {
	ApplicationRef,
	ComponentFactoryResolver,
	ComponentRef,
	Inject,
	Injectable,
	Injector,
	Type,
} from '@angular/core';

import { DIALOG_OVERLAY } from './dialog.conf';
import { IDialog } from './dialog.types';
import { IDialogActions, IDialogData } from './dialog.types';

@Injectable({
	providedIn: 'root',
})
export class DialogService {
	private dialogs: Set<ComponentRef<unknown>> = new Set<ComponentRef<unknown>>();
	private overlayRef: ComponentRef<unknown> = null;

	constructor(
		@Inject(DIALOG_OVERLAY) private overlay: Type<unknown>,
		private appRef: ApplicationRef,
		private resolver: ComponentFactoryResolver,
		private injector: Injector,
	) { }

	public openDialog({
		dialog,
		actions,
		data,
		hasDescription = true,
	}: {
		dialog: Type<IDialog>;
		actions?: IDialogActions;
		data?: IDialogData;
		hasDescription?: Boolean;
	}): ComponentRef<IDialog> {
		this.attachOverlay();

		const dialogRef = this.attachView(dialog);

		dialogRef.instance.actions = actions;
		dialogRef.instance.data = data;
		dialogRef.instance.closeDialog = this.bindDialogClose(dialogRef);
		dialogRef.instance.hasDescription = hasDescription;

		dialogRef.onDestroy(() => {
			this.detachView(dialogRef);
			this.dialogs.delete(dialogRef);

			if (this.dialogs.size === 0) {
				this.overlayRef.destroy();
			}
		});

		this.dialogs.add(dialogRef);

		this.attachRef(dialogRef, this.overlayRef);

		return dialogRef;
	}

	public closeDialog(dialogRef: ComponentRef<unknown>): void {
		if (!this.dialogs.has(dialogRef)) {
			return;
		}

		dialogRef.destroy();
	}

	public closeActiveDialog(): void {
		const lastAdded = Array.from(this.dialogs).pop();

		this.closeDialog(lastAdded);
	}

	private attachOverlay(): void {
		if (this.overlayRef) {
			return;
		}

		this.overlayRef = this.attachView(this.overlay);

		this.overlayRef.onDestroy(() => {
			this.detachView(this.overlayRef);
			this.overlayRef = null;
		});

		this.attachRef(this.overlayRef);
	}

	private detachView(typeRef: ComponentRef<unknown>): void {
		if (typeRef && typeRef.hostView) {
			this.appRef.detachView(typeRef.hostView);
		}
	}

	private attachView<T = unknown>(type: Type<T>): ComponentRef<T> {
		const typeFactory = this.resolver.resolveComponentFactory(type);
		const typeRef = typeFactory.create(this.injector);

		this.appRef.attachView(typeRef.hostView);

		return typeRef;
	}

	private attachRef(ref: ComponentRef<unknown>, root?: ComponentRef<unknown>): void {
		const rootElement = (root || this.appRef.components[0]).location.nativeElement;

		rootElement.appendChild(ref.location.nativeElement);
	}

	private bindDialogClose(ref: ComponentRef<IDialog>): () => void {
		const originalClose = ref.instance.closeDialog;

		return () => {
			if (originalClose) {
				originalClose.bind(ref.instance)();
			}

			this.closeDialog(ref);
		};
	}
}
