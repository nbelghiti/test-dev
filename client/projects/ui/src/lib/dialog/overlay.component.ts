import { Component, ElementRef, HostBinding, HostListener } from '@angular/core';

import { DialogService } from './dialog.service';

@Component({
	selector: 'ui-overlay',
	template: '',
	styleUrls: [
		'./overlay.component.scss',
	],
})
export class OverlayComponent {
	@HostBinding('class') public className = 'c-overlay';

	constructor(
		private dialogService: DialogService,
		private ref: ElementRef,
	) {}

	@HostListener('click', ['$event']) public onClick(e: MouseEvent) {
		const dialog = this.ref.nativeElement.querySelector('.c-dialog');

		if (dialog && (e.target === dialog || dialog.contains(e.target))) {
			return;
		}

		this.dialogService.closeActiveDialog();
	}
}
