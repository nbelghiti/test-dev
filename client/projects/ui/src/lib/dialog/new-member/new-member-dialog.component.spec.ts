
import { ComponentFixture, TestBed, async } from '@angular/core/testing';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { RouterTestingModule } from '@angular/router/testing';
import { TranslateModule } from '@ngx-translate/core';

import { ButtonComponent } from '../../button/button/button.component';
import { HeadingComponent } from '../../heading/heading.component';
import { IconComponent } from '../../icon/icon.component';
import { LinkComponent } from '../../link/link.component';
import { LoaderComponent } from '../../loader/loader.component';
import { SelectComponent } from '../../select/select.component';
import { DialogModule } from '../dialog.module';

import { NewMemberDialogComponent } from './new-member-dialog.component';

describe('NewMemberDialogComponent', () => {
	let component: NewMemberDialogComponent;
	let fixture: ComponentFixture<NewMemberDialogComponent>;

	beforeEach(async(() => {
		TestBed.configureTestingModule({
			imports: [
				RouterTestingModule,
				TranslateModule.forRoot(),
				ReactiveFormsModule,
				FormsModule,
				DialogModule
			],
			declarations: [
				NewMemberDialogComponent,
				IconComponent,
				LinkComponent,
				ButtonComponent,
				SelectComponent,
				LoaderComponent,
				HeadingComponent,
			],
		})
		.compileComponents();
	}));

	beforeEach(() => {
		fixture = TestBed.createComponent(NewMemberDialogComponent);
		component = fixture.componentInstance;
		fixture.detectChanges();
	});

	it('should create', () => {
		expect(component).toBeTruthy();
	});

	it('Should render correctly', () => {
		expect(fixture).toMatchSnapshot();
	});
});
