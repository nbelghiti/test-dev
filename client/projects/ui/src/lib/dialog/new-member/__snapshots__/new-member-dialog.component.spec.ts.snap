// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewMemberDialogComponent Should render correctly 1`] = `
<ui-new-member-dialog
  ButtonType={[Function Object]}
  HeadingLevel={[Function Object]}
  destroyed$={[Function Subject]}
  dialogClassName={[Function Boolean]}
  dialogOpened={[Function EventEmitter]}
  formBuilder={[Function FormBuilder]}
  newMemberForm={[Function FormGroup]}
  role={[Function String]}
>
  <div
    class="c-dialog__content"
  >
    <ui-heading
      classes="c-dialog__title"
      ng-reflect-classes="c-dialog__title"
      ng-reflect-level="2"
      ng-reflect-title="Add a new member."
      title="Add a new member."
    >
      <div
        class="c-heading c-dialog__title"
        ng-reflect-klass="c-heading c-dialog__title"
        ng-reflect-ng-class="[object Object]"
      >
        
        
        
        <h2>
          Add a new member.
        </h2>
        
        <div
          class="c-heading__meta"
        >
          
          
          
        </div>
      </div>
    </ui-heading>
    
    <form
      class="l-form__grid l-grid--with-spacing ng-untouched ng-pristine ng-invalid"
      ng-reflect-form="[object Object]"
      novalidate=""
    >
      <div
        class="l-grid__col--12"
      >
        <ui-select
          class="ng-untouched ng-pristine ng-invalid"
          name="role"
          ng-reflect-error="false"
          ng-reflect-label="NEW_MEMBER__TITLE_LABEL"
          ng-reflect-name="role"
        >
          <div
            class="c-select-field"
            ng-reflect-klass="c-select-field"
            ng-reflect-ng-class="[object Object]"
          >
            <select
              class="c-select-field__field ng-untouched ng-pristine ng-invalid"
              id="role"
              ng-reflect-name="role"
              ng-reflect-required=""
              required=""
            >
              <option
                disabled=""
                hidden=""
                ng-reflect-value=""
                selected=""
                value=""
              />
              
            </select>
            <label
              class="c-select-field__label"
              for="role"
            >
               NEW_MEMBER__TITLE_LABEL 
            </label>
            <div
              class="c-select-field__icon"
            >
              <ui-icon
                class="c-icon"
                ng-reflect-name="caret-down"
                ng-reflect-size="small"
              >
                <i
                  class="c-icon__glyph c-icon__glyph--small  fas fa-caret-down"
                />
              </ui-icon>
            </div>
            
          </div>
        </ui-select>
      </div>
    </form>
  </div><div
    class="c-dialog__actions"
  >
    <div
      class="c-dialog__action-item"
    >
      <ui-button
        ng-reflect-type="c-button--secondary"
      >
        
        <button
          class="c-button c-button--secondary"
          ng-reflect-klass="c-button c-button--secondary "
          ng-reflect-ng-class="[object Object]"
          title=""
        >
          
          
          <span
            class="c-button__text"
          >
            Cancel
          </span>
          <i
            class="c-button__loader fas fa-spinner fa-spin"
          />
          
          
        </button>
        
        
      </ui-button>
    </div>
    <div
      class="c-dialog__action-item"
    >
      <ui-button>
        
        <button
          class="c-button"
          ng-reflect-klass="c-button  "
          ng-reflect-ng-class="[object Object]"
          title=""
        >
          
          
          <span
            class="c-button__text"
          >
            Add team member
          </span>
          <i
            class="c-button__loader fas fa-spinner fa-spin"
          />
          
          
        </button>
        
        
      </ui-button>
    </div>
  </div>
</ui-new-member-dialog>
`;
